### PR TITLE
Strand-bias and codon fixing for ivar

### PR DESCRIPTION
**Pull request by saramonzon**
This PR is related to #247 
-  strand-bias filtering option is added to `ivar_variants_to_vcf.py` following instructions [here](https://gatk.broadinstitute.org/hc/en-us/articles/360035532152-Fisher-s-Exact-Test)
- Consecutive variants in ivar tsv output file are collapsed if they belong to the same codon, while parsing tsv file to vcf. 

## PR checklist

- [ X ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/viralrecon/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/viralrecon _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [X] Make sure your code lints (`nf-core lint`).
- [X] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: bin/ivar_variants_to_vcf.py [hunk 1]
```diff
@@ -28,27 +31,144 @@ def parse_args(args=None):
         default=0,
         help="Only output variants where allele frequency greater than this number (default: 0).",
     )
+    parser.add_argument(
+        "-nsb",
+        "--not_strand_bias",
```
Comment (drpatelh) [inline bin/ivar_variants_to_vcf.py:36]:
```suggestion
        "--ignore_strand_bias",
```

File: bin/ivar_variants_to_vcf.py [hunk 2]
```diff
@@ -28,27 +31,144 @@ def parse_args(args=None):
...
+        dest="NOT_STRAND_BIAS",
+        default=False,
+        help="Does not take into account strand bias, use this option when not using amplicons for sequencing",
+        action="store_true"
+    )
+    parser.add_argument(
+        "-nmc",
+        "--not_merge_codons",
```
Comment (drpatelh) [inline bin/ivar_variants_to_vcf.py:44]:
```suggestion
        "--ignore_codon_merging",
```

File: bin/ivar_variants_to_vcf.py (hunk 1 above)
Comment (saramonzon) [reply bin/ivar_variants_to_vcf.py:36]: ok!

File: bin/ivar_variants_to_vcf.py (hunk 2 above)
Comment (saramonzon) [reply bin/ivar_variants_to_vcf.py:44]: ok!

**Comment by saramonzon** [comment]
Ok, I will add the changes to the output there!
Test where failing due to the kraken db download...I re-run them, It should be fine now 🤞

**Review by drpatelh — COMMENTED** [review]
Thanks @saramonzon ! I had a look at the code yesterday and to be honest I got a bit lost and have no idea what is going on 😬 I am wondering whether we split out the scripts so we keep the `ivar_variants_to_vcf.py` script the same. And then have another script that specifically "cleans" the raw iVar output before running `ivar_variants_to_vct.py`? Quite a few people in the community are already using the `ivar_variants_to_vcf.py` script we wrote and I think we should keep it's functionality restricted to only converting the output files.

With regard to the codon implementation are you able to send all of the examples you are using to test this code (before and after processing)? I may have some ideas that could help with the implementation. One question I have already is would it help doing a single pass of the entire variants, reading them into an ordered dictionary (or other efficient data structure) and then processing the codons? I assume some of the code complexity comes from detecting the codons whilst traversing the file? Given that we are dealing with viral genomes I don't think it will be a big deal reading all of the variants into memory.

Let me know what you think and sorry to be a pain! I think the functionality is awesome but we need to think about how we support and maintain it in the future.

**Comment by saramonzon** [comment]
I think it's not necessary to split the scripts because we have the option set to keep the main functionality totally intact. We can split it if is mandatory but, maybe the other option is to set the parameters the opposite (`--collapse_codons` and `--strand_bias_filtering`) if you want the default to be the original code. This way it would be totally backwards compatible, just some new added functionality.

Regarding the codon implementation, I think the best example we have is the single-ended sample in the test-dataset, there are loads of consecutive variants and I've manually checked all of them and the code is working perfectly.

The complexity comes due to the same codon check, it's complicated at the biological side. The implementation is based on queues and stacks so only a 3 size list is always used, so it could make possible to parallel the code if this is used with many variants at any moment. We evaluated your option and you'll have to go through the dictionary or class equally once everything is read and make the same checks (check if consecutive, and check if same codon). Of course we could improve the readability, we are already working on that creating deque python library.

I understand your fear to support this functionality, but we are not going anywhere I hope :crossed_fingers:, we are totally using this functionality to all the genomics network for sars, and also for a bunch of other viruses. And as I've said the core functionality of the script is untouched so you can only not use the functionality and that's it!

**Comment by drpatelh** [comment]
Yep. The more I think about it the more I think that we need two separate scripts. One to filter the original iVar variant calls and one to convert them to VCF. It will also be more modular that way and when and "if" these issues are fixed on the iVar side we can just remove the filtering options.

I am happy to merge this PR in but I would like to have a play with it at some point to understand what is going on and maybe tweak some stuff - mainly for my benefit :)

**Review by drpatelh — APPROVED** [review]

**Comment by drpatelh** [comment]
Love that you guys really look into the small print and find all of these issues! 👌🏽 🤩

**Comment by drpatelh** [comment]
Feel free to merge when you are happy @saramonzon !

**Comment by saramonzon** [comment]
Thanks very much! we will try and separate it in two scripts! I will merge this and do that on a separate PR, it currently works this way!
I hope when that issues get fixed, ivar will also provide a vcf format so neither of them are necessary!

**Pull request merged**